docs/DesignDocs/SealingAPI.md

```diff
@@ -0,0 +1,591 @@
+                              OE_SEAL_SGX | OE_SEALKEY_DEFAULT_FLAGSMASK |
+                              SGX_FLAGS_PROVISION_KEY,
+                              (uint8_t**)&key_info, &key_info_size);
+if (result != OE_OK)
+    // Handle errors...
+else
+    key_info->key_name = SGX_KEYSELECT_PROVISION_SEAL;
+```
+
+## Specification
+
+New files:
+
+- [include/openenclave/bits/seal.h](../../include/openenclave/bits/seal.h)
+  contains generic definitions.
+- [include/openenclave/bits/sgx/seal.h](../../include/openenclave/bits/sgx/seal.h)
+  contains SGX specific definitions.
+- [enclave/seal_gcm-aes.c](../../enclave/seal_gcm-aes.c) implements `oe_seal()`
+  and `oe_unseal()`.
+- [enclave/crypto/gcm.c](../../enclave/crypto/gcm.c) implements GCM-AES128.
+
+Modified files:
+
+- [include/openenclave/enclave.h](../../include/openenclave/enclave.h)
+  declares `oe_get_seal_key_info()`, `oe_seal()` and `oe_unseal()`.
+- [enclave/core/sgx/keys.c](../../enclave/core/sgx/keys.c) implements
+  `oe_get_seal_key_info()`.
+
+### oe_seal()
+
+```C
+/**
+ * Seal data to an enclave using AEAD (Authenticated Encryption with Additional
+ * Data).
+ *
+ * @param[in] key_info The enclave-specific key information to derive the seal
+ * key with.
+ * @param[in] key_info_size The size of the \p key_info buffer.
+ * @param[in] plaintext Optional buffer to be encrypted under the seal key.
+ * @param[in] plaintext_size Size of \p plaintext, must be \c 0 if \p plaintext
+ * is \c NULL.
+ * @param[in] additional_data Optional additional data to be included in the
+ * final MAC.
+ * @param[in] additional_data_size Size of \p additional_data, must be \c 0 if
+ * \p additional_data is \c NULL.
+ * @param[out] blob On success, receive the pointer to a buffer containing both
+ * \p additional_data and encrypted \p plaintext, along with necessary
+ * information for unsealing. Freed by oe_free().
+ * @param[out] blob_size On success, receive the size of \p blob.
+ *
+ * @retval OE_OK \p plaintext and \p additional_data were sealed to the enclave
+ * successfully.
+ * @retval OE_INVALID_PARAMETER At least one parameter is invalid.
+ * @retval OE_OUT_OF_MEMORY Failed to allocate memory.
+ * @retval OE_CRYPTO_ERROR Error occurred during encryption.
+ */
+oe_result_t oe_seal(
+    const uint8_t* key_info,
+    size_t key_info_size,
+    const uint8_t* plaintext,
+    size_t plaintext_size,
+    const uint8_t* additional_data,
+    size_t additional_data_size,
+    uint8_t **blob,
+    size_t *blob_size);
+```
+
+`oe_seal()` encrypts and packages user data into a sealed blob using *AEAD*
+(*Authenticated Encryption with Additional Data*). A sealed blob is comprised
+of a header, followed immediately by payload, which is the concatenation of
+cipher text and *AAD* (*Additional Authenticated Data*).
+
+The blob header is TEE specific. On SGX, for interoperability, we adopted the
+same format as the Intel SGX SDK. Below gives the definition of
+`oe_sealed_blob_header_t` on SGX.
```

Comment:
```suggestion
`oe_sgx_sealed_blob_header_t` on SGX.
```

Reply:
`oe_sealed_blob_header_t` isn't SGX specific. It's specific to `oe_seal()` implementation instead. Let me move it to a TEE neutral header file.

Reply:
I take my words back. `oe_sealed_blob_header_t` is SGX specific because of `key_info`. But if treating `key_info` as just an octet string, then the `oe_seal()` implementation is TEE agnostic.  

How about the following?

In `bits/sgx/seal.h`:

```C
typedef sgx_key_request_t _oe_seal_key_info_t;
```

Btw, what's the naming convention for *internal* structs?

In `bits/seal.h`:

```C
typedef {
    _oe_seal_key_info_t key_info;
    // ...
} oe_sealed_blob_header_t;
```

Given the header is cipher/mode specific, how about renaming it to `oe_sealed_blob_header_gcm_aes`?

Reply:
If it's SGX specific, then the sgx-specific struct should have sgx in the name.  It can be passed in a TEE-agnostic uint8_t* buffer argument for example.

Reply:
The bad thing about union is errors can't be caught at compile time. The structure `sgx_key_request_t` won't be defined when the target TEE isn't SGX, so union won't work anyway. What we are trying to do here, using C++ terms, is to instantiate a template, with the argument `_oe_seal_key_info_t` set to `sgx_key_request_t`. Would this `typedef` below be an acceptable solution?

```C
typedef sgx_key_request_t _oe_seal_key_info_t;
```

If not, is C++ allowed in OE SDK? Shall I use C++ template instead?

Reply:
we need APIs to be usable by C.  (Maybe there's C++ in addition, but you must be able to seal/unseal from C)